Make 'retry' property on errors from checkoutAfterProcessingWithSuccess/Error observers default to true if it's undefined

Aljullu opened the pull request:
Fixes #3221.

Errors in WC Payments were not recoverable because the `retry` property was not specified in the error response, so it was considered `false`. This behavior was a bit inconsistent in our codebase, because `retry` [needed to be truthy in `CHECKOUT_AFTER_PROCESSING_WITH_SUCCESS`](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/assets/js/base/context/cart-checkout/checkout-state/index.js#L281), but [was allowed to be undefined in `CHECKOUT_AFTER_PROCESSING_WITH_ERROR`](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/assets/js/base/context/cart-checkout/checkout-state/index.js#L240-L243).

This PR updates the Checkout Provider so if retry is undefined, it treats it as if it was `true`.

### How to test the changes in this Pull Request:

* Install WC Payments and set it up.
* Submit an order using WC Payments with the card number `4000 0027 6000 3184`, but in the validation popup, press on `Fail authentication` or `Cancel`.
* Verify the `Place Order` button doesn't stay disabled and you can resubmit the order.

### Changelog

> Dev: errors produced on `checkoutAfterProcessingWithSuccess/Error` observers no longer need `retry: true` to allow the user to retry the payment. You can set `retry: false` if you want to force disabling submitting the checkout form again. In addition, a new `shouldRetry` function is available from the `useEmitResponse` hook which enables setting more complex retry rules.

senadir commented:
feels weird to treat undefined as true, why not make sure payment methods return the correct value?

What's the behavior we're aiming for here?

- Always retry unless specified
- Never retry unless specified

senadir commented:
Maybe, based on that, we can decide to rename this variable to make it obvious.

Aljullu commented:
> why not make sure payment methods return the correct value?

The idea is to provide good defaults, so there is less risk for payment methods to introduce bugs (like #3221).

I don't have any data supporting why the default should be true instead of false, so I'm happy to be proven wrong, but at least for Stripe and WC Payments we needed to support retrying after an error, and I can imagine that's the most common behavior among other payment methods.

In any case, `checkoutAfterProcessingWithSuccess` and `checkoutAfterProcessingWithError` are currently behaving differently, so whatever decision we end up taking, we need to make them work consistently.

> Always retry unless specified

Correct, that's the behavior we are trying to achieve: always retry unless `retry: false` (see discussion in 978-GH-woocommerce-payments).

Note: If we were to rename `retry` to something else, that would probably be a breaking change.

senadir reviewed (approved):
I think this makes sense in the broader terms, maybe we can change the wording from 'undefined' to saying it defaults to true. 

I'd probably also wait and see what Darren has to say about this.

nerrad reviewed (approved):
Looks good to me, I do wonder if we should add a `shouldRetry` function to the `useEmitResponse` hook to include this check and logic centralized there? That way the condition could just be:

```js
if ( shouldRetry( response ) )
```

Which is a bit clear and keeps the actual default coercion logic in the `useEmitResponse` hook?

I'm going to pre approve anyways - up to you Albert whether you modify for the above or not.

Aljullu commented:
Thanks for the reviews @senadir @nerrad!

I moved the logic of `shouldRetry` to `useEmitResponse` in 7fa0582fda1edb7de8f9b39ead3696f7de3f45c2 and updated the changelog notice of this PR accordingly.
